### PR TITLE
[CSL-1765] automate update system proposals

### PR DESCRIPTION
**Pull request by pva701**
1. Added `primary` flag for `add-key`
2. Added `vote-all` flag for `propose-update`
To get more information see descriptions of commits.

### CODE REVIEW

File: auxx/src/Command/Types.hs [hunk 1]
```diff
@@ -52,7 +51,7 @@ data Command
     | Send Int (NonEmpty TxOut)
     | SendToAllGenesis !SendToAllGenesisParams
     | Vote Int Bool UpId
-    | ProposeUpdate !ProposeUpdateParams
+    | ProposeUpdate !Int !ProposeUpdateParams !Bool
```
Comment (gromakovsky) [inline auxx/src/Command/Types.hs]: Why are these values (`Int` and `Bool`) not in `ProposeUpdateParams`?

File: auxx/src/Command/Update.hs [hunk 2]
```diff
@@ -127,10 +131,34 @@ propose sendActions ProposeUpdateParams{..} = do
...
+        let safeZip a b =
+                if length a == length b then zip a b
+                else error $ "Number of signatures: " <> show (length b) <> ", expected " <> show (length a)
+        forM (safeZip skeys signatures) $ \(skey, signature) -> pure $
+            UpdateVote
+                { uvKey        = encToPublic skey
+                , uvProposalId = upid
+                , uvDecision   = True
+                , uvSignature  = signature
+                }
```
Comment (gromakovsky) [inline auxx/src/Command/Update.hs]: I think the code would be cleaner if you changed the second argument of `submitUpdateProposal` to `[SafeSigner]` (i. e. generalize it to multiple signers) and implemented the `not voteAll` case using that generalized version.

File: auxx/src/Command/Types.hs (hunk 1 above)
Comment (pva701) [reply auxx/src/Command/Types.hs]: I assumed `ProposeUpdateParams` contains related to `UpdateProposal` fields, like software version and etc.

**Comment by pva701** [comment]
>You forgot to remove Rearranging the keys step from the document.
In which document? I have no clue what is going on in documents and which documents exist

**Comment by gromakovsky** [comment]
> In which document?

`docs/how-to/push-update-on-mainnet.md`, that's the main document used by DevOps to send updates to mainnet.

File: auxx/src/Command/Types.hs (hunk 1 above)
Comment (gromakovsky) [reply auxx/src/Command/Types.hs]: Not really, `ProposeUpdateParams` contains all parameter for `ProposeUpdate` command. That's what the documentation says.

Comment (gromakovsky) [reply auxx/src/Command/Types.hs]: Also it's better to do as little changes as possible in release branches, because otherwise merging will be harder.

Comment (pva701) [reply auxx/src/Command/Types.hs]: >merging will be harder.
pffff, who cares :yao:

File: auxx/src/Command/Run.hs [hunk 3]
```diff
@@ -95,7 +96,7 @@ runCmd sendActions (SendToAllGenesis stagp) =
     Tx.sendToAllGenesis sendActions stagp
 runCmd sendActions (Vote idx decision upId) =
     Update.vote sendActions idx decision upId
-runCmd sendActions (ProposeUpdate params) =
+runCmd sendActions (ProposeUpdate params)  =
```
Comment (gromakovsky) [inline auxx/src/Command/Run.hs]: _mda_

File: auxx/src/Command/Update.hs [hunk 4]
```diff
@@ -113,24 +113,31 @@ propose sendActions ProposeUpdateParams{..} = do
...
-                    let id = hash updateProposal
-                    putText $
-                      sformat ("Update proposal submitted, upId: "%hashHexF) id
+
+    skeys <- if not puVoteAll then pure [skey]
+             else getSecretKeysPlain
+    withSafeSigners skeys (pure emptyPassphrase) $ \ss -> do
+        unless (length skeys == length ss) $
+            error $ "Number of safe signers: " <> show (length ss) <> ", expected " <> show (length skeys)
+        let publisherSS = ss !! if not puVoteAll then 0 else puIdx
```
Comment (gromakovsky) [inline auxx/src/Command/Update.hs]: AFAIU, `publisherSS` is the same as `skey`. Do we need two different variables?

File: docs/how-to/push-update-on-mainnet.md [hunk 5]
```diff
@@ -69,42 +69,35 @@ COMMONOPTS="--system-start 0 --configuration-file ../cardano-sl/node/configurati
 AUXXOPTS="--log-config ../cardano-sl/scripts/log-templates/log-config-qa.yaml --logs-prefix logs/aux-update.1.0.1 --db-path aux-update-1.0.1 --peer ${RELAY_PEER}"
 ```
 
-Rearranging the keys
---------------------
+Sending an update proposal
+------------------
 
-Apply `cardano-keygen rearrange` to the keys to make them usable for proposing an update and voting:
+Let's say that you want to push an update with the following Windows/macOS installers (note: it's important that the installers might be in current directory, i.e. there must be no slashes):
```
Comment (gromakovsky) [inline docs/how-to/push-update-on-mainnet.md]: `installers might be` or `installers must be`? 🤔

File: docs/how-to/push-update-on-mainnet.md [hunk 6]
```diff
@@ -69,42 +69,35 @@ COMMONOPTS="--system-start 0 --configuration-file ../cardano-sl/node/configurati
...
 
-Importing the keys
-------------------
-
-Import secret keys: only 4 of 7 is needed (extra votes will be ignored). Cluster nodes all have equal stake and more than a half stake's votes is needed to make a decision about update (approve/dismiss). Imported keys are stored locally in `secret.key` which you **must** delete after completing this guide (unless there's a good valid reason not to).
+First of all you have to import secret keys. Import secret keys: only 4 of 7 is needed (extra votes will be ignored). Cluster nodes all have equal stake and more than a half stake's votes is needed to make a decision about update (approve/dismiss). Imported keys are stored locally in `secret.key` which you **must** delete after completing this guide (unless there's a good valid reason not to).
 
 ```
-stack exec -- cardano-auxx $COMMONOPTS $AUXXOPTS cmd --commands "add-key key0.sk, add-key key1.sk, add-key key2.sk, add-key key3.sk, listaddr"
+stack exec -- cardano-auxx $COMMONOPTS $AUXXOPTS cmd --commands "add-key key0.sk primary, add-key key1.sk primary, add-key key2.sk primary, add-key key3.sk primary, listaddr"
```
Comment (gromakovsky) [inline docs/how-to/push-update-on-mainnet.md]: I would write what `primary` means.

**Comment by gromakovsky** [comment]
Did you test it locally?

File: auxx/src/Command/Update.hs (hunk 4 above)
Comment (pva701) [reply auxx/src/Command/Update.hs]: lol `skey` is secret key, publisherSS is safe signer

**Comment by pva701** [comment]
didn't test, users will test

File: node/src/node/Main.hs [hunk 7]
```diff
@@ -75,3 +75,4 @@ main = do
     CLI.printFlags
     let conf = CLI.configurationOptions (CLI.commonArgs commonNodeArgs)
     runProduction $ withConfigurations conf (action args)
+
```
Comment (gromakovsky) [inline node/src/node/Main.hs]: :cho:

Comment (pva701) [reply node/src/node/Main.hs]: niche

fucking stack doesn't osil to re-link our fucking binaries without changing them

Comment (gromakovsky) [reply node/src/node/Main.hs]: Clean 🤷‍♂️

Comment (pva701) [reply node/src/node/Main.hs]: I didn't want to trigger CI because of it, so thx for approve

Comment (gromakovsky) [reply node/src/node/Main.hs]: But CI failed anyway 🤷‍♂️

**Pull request merged**